### PR TITLE
[quant][graphmode][fx] Add reference option support for linear_dynamic_fp16

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52865 [quant][graphmode][fx][fp16] Add fp16 support for silu
* #52864 [quant][graphmode][fx][fp16] Add fp16 support for tanh
* #52863 [quant][graphmode][fx][fp16] Add fp16 support for sigmoid
* #52862 [quant][graphmode][fx][fp16] Add fp16 support for layer_norm
* #52811 [quant][graphmode][fx][fp16] Add fp16 support for torch.sum
* #52810 [quant][graphmode][fx][fp16] Add fp16 support for div
* #52809 [quant][graphmode][fx][fp16] Add fp16 support for sub
* #52808 [quant][graphmode][fx] Add support for fp16 bmm pattern
* #52807 [quant][graphmode][fx][test][refactor] Refactoring binary op tests to split int8 and float16 tests
* #52714 [quant][graphmode][fx][fp16] Add fp16 support for {add|mul}{_relu}
* #52651 [quant][refactor] Merge add and mul handler
* #52650 [quant][graphmode][fx] Add reference option support for linear_static_fp16
* **#52534 [quant][graphmode][fx] Add reference option support for linear_dynamic_fp16**

Summary:
Currently linear_dynamic_fp16 has a signature that's tied to fbgemm/qnnpack
We'll need to produce a pattern equivalent to linear_dynamic_fp16 to support extensions
to other backends

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_linear_dynamic_fp16

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26557726](https://our.internmc.facebook.com/intern/diff/D26557726)

### CODE REVIEW

File: torch/quantization/fx/utils.py [hunk 1]
```diff
@@ -94,23 +94,30 @@ def get_per_tensor_qparams(activation_post_process):
     dtype = activation_post_process.dtype
     return scale, zero_point, dtype
 
-def get_quantize_op_and_qparams(activation_post_process):
+def get_quantize_node_info(activation_post_process):
```
Comment (vkuzo) [inline torch/quantization/fx/utils.py:97]: any chance we can annotate the return type?  Would be great for readability, especially since we are changing the return type

File: torch/quantization/qconfig.py [hunk 2]
```diff
@@ -64,7 +64,7 @@ def __new__(cls, activation=torch.nn.Identity, weight=torch.nn.Identity):
 
 default_dynamic_qconfig = QConfigDynamic(activation=default_dynamic_quant_observer,
                                          weight=default_weight_observer)
-float16_dynamic_qconfig = QConfigDynamic(activation=PlaceholderObserver.with_args(dtype=torch.float16),
+float16_dynamic_qconfig = QConfigDynamic(activation=PlaceholderObserver.with_args(dtype=torch.float32),
```
Comment (vkuzo) [inline torch/quantization/qconfig.py:67]: just curious, what's the context for this change?

**Review by vkuzo — APPROVED** [review]

File: torch/quantization/qconfig.py (hunk 2 above)
Comment (jerryzh168) [reply torch/quantization/qconfig.py:67]: this is incorrect previously I think, float16_dynamic quant is taking fp32 activation as input before

File: torch/quantization/fx/utils.py (hunk 1 above)
Comment (jerryzh168) [reply torch/quantization/fx/utils.py:97]: sure

File: torch/quantization/fx/utils.py [hunk 3]
```diff
@@ -94,23 +94,30 @@ def get_per_tensor_qparams(activation_post_process):
...
+        if is_per_channel(activation_post_process.qscheme):
+            ch_axis = int(activation_post_process.ch_axis)
+            qparams = {"_scale_": scale, "_zero_point_": zero_point, "_axis_": ch_axis, "_dtype_": dtype}
+            quantize_op = torch.quantize_per_channel
+        else:
+            scale = float(scale)
+            zero_point = int(zero_point)
+            qparams = {"_scale_": scale, "_zero_point_": zero_point, "_dtype_": dtype}
+            quantize_op = torch.quantize_per_tensor  # type: ignore
+    elif dtype == torch.float16:
```
Comment (vkuzo) [inline torch/quantization/fx/utils.py:116]: this is for my own learning, some other places in this PR change the dtype on the observers from float16 to float32, where is this dtype==torch.float16 coming from?

Comment (jerryzh168) [reply torch/quantization/fx/utils.py:116]: this is for supporting generating patterns for quantizing to fp16 precision (change of fp16 to fp32 is a fix that's unrelated to this check)
